flow.go

```diff
@@ -0,0 +1,542 @@
+				if !frontmatterRemoved && len(prebuffer) > 0 {
+					// EOF reached while detecting frontmatter
+					// Treat remaining prebuffer as content
+					fb.accum = append(fb.accum, prebuffer...)
+				}
+
+				// Final flush of all remaining content
+				if len(fb.accum) > 0 {
+					return fb.flush(true)
+				}
+				// Empty input special case - glamour outputs \n\n for empty input
+				if fb.written == 0 {
+					// Never had any output, render empty string to match glamour
+					return fb.render([]byte{}, true)
+				}
+				return nil
+			}
+
+			if result.err != nil {
+				return result.err
+			}
+		}
+	}
+}
+
+// shouldFlush determines if buffer should be flushed based on flow config
+func (fb *FlowBuffer) shouldFlush() bool {
+	if fb.config.Window == FlowNoWindow {
+		// No flow: never flush until EOF
+		return false
+	}
+	if fb.config.Window < FlowNoWindow {
+		// Aggressive flow: flush at every safe boundary
+		return fb.hasSafeBoundary()
+	}
+	// Bounded flow: flush when exceeding size threshold
+	return int64(len(fb.accum)) >= fb.config.Window
+}
+
+// findFrontmatterEnd finds the end of frontmatter in buffer, returns index after frontmatter
+func (fb *FlowBuffer) findFrontmatterEnd(data []byte) int {
+	// Already checked that data starts with ---
+	// Look for closing --- on its own line
+
+	// Skip the opening ---
+	idx := 3
+	if idx < len(data) && data[idx] == '\r' {
+		idx++
+	}
+	if idx < len(data) && data[idx] == '\n' {
+		idx++
+	}
+
+	// Search for closing ---
+	for idx < len(data) {
+		// Look for \n--- sequence
+		if idx > 0 && data[idx-1] == '\n' {
+			// Check if we have ---
+			if idx+3 <= len(data) && bytes.Equal(data[idx:idx+3], []byte("---")) {
+				// Check if followed by newline or EOF
+				endIdx := idx + 3
+				if endIdx >= len(data) {
+					return endIdx // EOF after ---
+				}
+				if data[endIdx] == '\n' {
+					return endIdx + 1
+				}
+				if data[endIdx] == '\r' && endIdx+1 < len(data) && data[endIdx+1] == '\n' {
+					return endIdx + 2
+				}
+			}
+		}
+		idx++
+	}
+
+	return -1 // No closing found
+}
+
+// hasSafeBoundary checks if buffer contains a complete block
+func (fb *FlowBuffer) hasSafeBoundary() bool {
+	// Look for block boundaries: empty lines, code fences, headers
+	// Check for complete code block using level-aware detection
+	if bytes.Contains(fb.accum, []byte("```")) {
+		// Use the level-aware boundary finder
+		if fb.findCodeBlockBoundary(fb.accum) > 0 {
+			return true
+		}
+	}
+
+	// Check for paragraph boundary (empty line)
+	if bytes.Contains(fb.accum, []byte("\n\n")) {
+		return true
+	}
+
+	// Check for header followed by content
+	lines := bytes.Split(fb.accum, []byte("\n"))
+	for i, line := range lines {
+		if bytes.HasPrefix(line, []byte("#")) && i < len(lines)-1 {
+			// Header with at least one line after it
+			return true
+		}
+	}
+
+	return false
+}
+
+// findSafeBoundary finds the last safe split point in the buffer
+func (fb *FlowBuffer) findSafeBoundary() int {
```

Comment:
We probably want to avoid searching the entire accumulated buffer every time, even though it's kind of an edge case. Since we don't yet consider the internals of tables or code blocks to contain safe boundaries, it's far **more likely** to accumulate larger buffers, but still only really becomes a problem if they become pathologically large. At any rate, keeping track of the already-searched offset is probably straightforward and worth looking at before landing.

Reply:
This is done